docs/index.rst

```diff
@@ -3,8 +3,128 @@
+reasoning behind some of these philosophies.
+
+Modularity
+^^^^^^^^^^
+
+``uarray`` (and its sister modules ``unumpy`` and others to come) were designed
+from the ground-up to be modular. This is part of why ``uarray`` itself holds
+the core backend and dispatch machinery, and ``unumpy`` holds the actual
+multimethods. Also, ``unumpy`` can be developed completely separately to
+``uarray``, although the ideal place to have it would be NumPy itself.
+
+However, the benefit to having it separate is that it could span multiple
+NumPy versions, even before NEP-18 (or even NEP-13) was available.
+
+Separate Imports
+^^^^^^^^^^^^^^^^
+
+Code wishing to use the backend machinery for NumPy (as an example) will
+use the statement ``import unumpy as np`` instead of the usual
+``import numpy as np``. This is deliberate: it makes dispatching opt-in
+instead of being forced to use it, and the overhead associated with it.
+However, a package is free to define its main methods as the dispatchable
+versions, thereby allowing dispatch on the default implementation.
+
+Extensibility *and* Choice
+^^^^^^^^^^^^^^^^^^^^^^^^^^
+
+If some effort is put into the dispatch machinery, it's possible to
+dispatch over arbitrary objects --- including arrays, dtypes, and
+so on. A method defines the type of each dispatchable argument, and
+backends are *only* passed types they know how to dispatch over.
+
+Methods can have a default implementation in terms of other methods,
+but they're still overridable.
+
+This means that only one framework is needed to, for example, dispatch
+over ``ufunc`` s, arrays, dtypes and all other primitive objects in NumPy,
+while keeping the core ``uarray`` code independent of NumPy and even ``unumpy``.
+
+Backends can span modules, so SciPy could jump in and define its own methods
+on NumPy objects and make them overridable within the NumPy backend.
+
+User Choice
+^^^^^^^^^^^
+
+The users of ``unumpy`` or ``uarray`` can choose which backend they want
+to prefer with a simple context manager. They also have the ability to
+force a backend, and to skip a backend. This is useful for array-like
+objects that provide other array-like objects by composing them. For
+example, Dask could perform all its blockwise function calls with the
+following psuedocode (obviously, this is simplified):
+
+.. code:: python
+
+    in_arrays = extract_inner_arrays(input_arrays)
+    out_arrays = []
+    for input_arrays_single in in_arrays:
+        args, kwargs = blockwise_function.replace_args_kwargs(
+            args, kwargs, input_arrays_single)
+        with ua.skip_backend(DaskBackend):
+            out_arrays_single = blockwise_function(*args, **kwargs)
+        out_arrays.append(out_arrays_single)
+
+    return combine_arrays(out_arrays)
+
+A user would simply do the following:
+
+.. code:: python
+
+    x = da.array([a, b, c])
+
+    # Anything that uses x will (usually) use the Dask backend
+
+    with ua.use_backend(DaskBackend):
```

Comment:
The combination of these examples makes me wonder what happen if a library skips a backend that the user forces. Does it fall back to a default. Does there *have to be* a default?

Reply:
There's no default. In the current code, it will be skipped. The skip is really just meant to be a way to avoid infinite recursion for libraries that both consume and export backends with intersecting method spaces.

Reply:
To be more explicit: To `uarray`, NumPy is just a backend. There's no default at all, it'll just error if you don't select one. It's easy to add the capability for a default. You can provide a default implementation for methods (perhaps even in terms of others), but it can be overridden.